### ISSUE TITLE
What is the structure and contents of the subfolders of the proprocessed directory?

### ISSUE THREAD

**Issue by frederikfaye** [MIC-DKFZ/nnUNet#143]
After running the data preparation (for the Hippocampus dataset), I end up with a several subfolders in the nnUNet_preprocessed/Task* folder: 
- nnUNet_stage0 (containing a .npz and a .pkl file for each of the 260 unique IDs)
- nnUNet_2D_stage0 (containing a .npy, .npz and a .pkl file for each of the 260 unique IDs)
- gt_segmentations (containing a .nii.gz file for each of the 260 unique IDs)

I looked into the nnUNet_2D_stage0 .npy files, and they have been normalized (and I guess cropped, although it doesn't seem to anything for this dataset, which is fine), but not interpolated, so each image has different dimensions. 
Also, for each ID (i.e. the patient ID), the .npy file has had an extra dimension put into it, and two of the axes have been swapped, such that e.g. `shape(X_raw)=(34,52,35)` becomes `shape(X_preprocessed)=(2,35,42,34)`. It seems that `X_preprocessed[0,:,:,:]`is the image, while `X_preprocessed[1,:,:,:]` is the segmentation groundtruth, but I'm not sure, and I would like to be sure. 

This leads me to the following list of questions:

1. Which files are duplicates, and which are the ones I should be using for training and validation? Can I ignore the non-.npy files?
2. What is the structure of the each of the filetypes which are not duplicates?
3. Why hasn't the data been interpolated? Are all the voxels already positioned completely isotropically for the Hippocampus dataset?
4. How can I get the data interpolated along the z-axis (the "slices axis"), such that the images for each ID has the same number of slices along z?
5. As far as I have understood, the training set contains 260 datapoints, corresponding to the contents of the subfolders - but where is the test data?

Thank you very much for this repository, and for your help!

**Comment by FabianIsensee** [comment]
Hi,
I believe you are misinterpreting what resampling means. The data in the npz and npy files is already resampled. Resampling here means that they have the same voxel spacing, not the same size! Note that all the Hippocampus data is already in 1x1x1 isotropic spacing, so there is no resampling happening for that dataset.

nnUNet_stage0 contains the preprocessed (cropped, normalized, resampled) data for the 3d_fullres U-net

nnUNet_2D_stage0 contains the preprocessed (cropped, normalized, resampled) data for the 2d U-net

The disticntion between 2d and 3d unet is necessary because the resampling for the 2d u-net happens in 2d only (the z axis is unchanged) while the data for the 3d unet is resampled along all axes.

The npz files are just containers for the npy files. Thus they contain the same information. This has something to do with how our GPU cluster works. For the training only the npy files are needed, but it could be that some parts of the pipeline still rely on the npz files to be present so I strongly recommend you leave them as they are. The pkl files store the geometry information and are absolutely required.

> Which files are duplicates, and which are the ones I should be using for training and validation? Can I ignore the non-.npy files?

You should let nnU-net handle the splitting for training and validation. It does that automatically.

> What is the structure of the each of the filetypes which are not duplicates?

pkl contains geometry, npz and npy contain the data. The shape is (c, z, x, y) where c is the number of modalities + 1. The last image in the 4d stack is always the gt segmentation

> Why hasn't the data been interpolated? Are all the voxels already positioned completely isotropically for the Hippocampus dataset?

This data is interpolated. Do not confuse shape with spacing! Hippocampus is a corner case where all images already have the same spacing (this is not the case for most other datasets) so the data here did not need resampling

> How can I get the data interpolated along the z-axis (the "slices axis"), such that the images for each ID has the same number of slices along z?

Why would you want that? That could distort the images. Image shapes are irrelevant in the medical domain - the voxel spacing needs to be the same throughout the dataset

> As far as I have understood, the training set contains 260 datapoints, corresponding to the contents of the subfolders - but where is the test data?

The test data will be preprocessed on the fly when you run the inference script. No need to do this now.

Hope this helps.
Best,
Fabian

**Comment by frederikfaye** [comment]
Thanks for your quick answer!

I am still not sure what the gt_segmentations folder contain. The same as the gt segmentations in the .npy files? 

> The shape is (c, z, x, y) where c is the number of modalities + 1. The last image in the 4d stack is always the gt segmentation

What do you mean by "the last"? If the shape of .npy file containing `X` is `(c, z, x, y)`, would the gt segmentation then be `X[-1,:,:,:]`, consisting of values in `{0,1,..,L}`, where `L+1` is the number of different labels (if all different labels are present in this X)?

> This data is interpolated [...] so the data here did not need resampling

If there is no resampling, then there have been no interpolation carried out, right?

Thanks, 
Frederik

**Comment by FabianIsensee** [comment]
> I am still not sure what the gt_segmentations folder contain. The same as the gt segmentations in the .npy files?

They are the ground truth segmentations. As you may have noticed those are .nii.gz. They are just a copy of the reference segmentations.

> If the shape of .npy file containing X is (c, z, x, y), would the gt segmentation then be X[-1,:,:,:], consisting of values in {0,1,..,L}, where L+1 is the number of different labels (if all different labels are present in this X)?

correct. But there is also a -1 in there to indicate where all modalities were 0. This -1 is most of the time unused though

> If there is no resampling, then there have been no interpolation carried out, right?

For hippocampus there is no resampling = no interpolation. For practically all datasets interpolation is carried out, though.

Best,
Fabian

**Comment by frederikfaye** [comment]
Thanks again for your answers, they are very helpful! 

> But there is also a -1 in there to indicate where all modalities were 0. This -1 is most of the time unused though

Why would the dataset curators not just set all pixel values to zero instead? Would it be safe to set all these -1's (in the segmentation map) to 0's instead? This has not necessarily anything to do with how this would interact with nnUNet - I'm simply asking in order to understand the data (and the preprocessing). 

Best,
Frederik

**Comment by FabianIsensee** [comment]
Hi,
these -1 are created by nnU-net to indicate where the images were all 0. This is used for cropping and for some datasets also during normalization. This is something specific to nnU-net and I don't think this is standard. 
Here is a quick summary
- nnU-net checks where all images are 0. This sometimes indicated background (for example brats). nnU-net then crops to nonzero region in the image
- in the remaining (cropped) image, the nonzero mask is retained. This is done by encoding it into the segmentation map (-1)
- if the cropping resulted in a substantial reduction in image size, the normalization (z-score) takes place within the nonzero mask only. Again, this replicates standard preprocessing for brats-like data

For all datasets where cropping did not do anything you can just ignore the -1. This is exactly what nnU-net does as well

**Comment by frederikfaye** [comment]
I'm not sure I understand the implementation of this: I have found several examples of two or three pixels that were -1 in segmentation maps where neither the input image, nor the rest of the segmentation map were all 0. Is this intended?

**Comment by frederikfaye** [comment]
Should I make a separate issue about the "stray -1s" problem?

**Comment by FabianIsensee** [comment]
Interesting. You can leave this here.
Can you please provide examples? Ideally from the MSD?

**Comment by frederikfaye** [comment]
Something as simple as the following shows the -1s in the segmentation ground truths:

```python
import numpy as np
from glob import glob

data_dir = './nnUNet_base/nnUNet_preprocessed/Task04_Hippocampus/nnUNet_2D_stage0/'
img_paths = glob(data_dir + '*.npy')

img_paths

for img_path in img_paths:
    img = np.load(img_path)
    print(np.unique(img[1,:,:,:]))
```

**Comment by FabianIsensee** [comment]
Hi Frederik,
I will have a look at this, thanks!
Best,
Fabian
